Implement support for vllm as alternative backend (#3415)
kserve/kserve: python/huggingfaceserver/README.md

      - --model_id=bert-base-uncased
      - --predictor_protocol=v2
      - --tensor_input_names=input_ids

>       - --disable_vllm

Reviewer: I think we probably we do not want to confuse user that they have to set this config to get Bert to work. I think a better example would be the case that vllm is by default and show an example that it can be disabled.

Author: I have added a disable vllm example

Reviewer 2: I just trying out the new huggingface runtime and notice this. I feel like a bit werid to have vllm as default runtime (as the PR mentioned vllm as alternative backend).

how about `--backend=vllm`, having `huggingface` as default and vllm as alternative and future can support other backend like `tensorRT`

what do you think? @yuzisun

apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: huggingface
spec:
  predictor:
    containers:
    - args:
      - --model_name=llama2
      - --model_id=meta-llama/Llama-2-7b-chat-hf
      image: kserve/huggingfaceserver:latest
      name: kserve-container
      resources:
        limits:
          cpu: "6"
          memory: 24Gi
          nvidia.com/gpu: "1"
        requests:
          cpu: "6"
          memory: 24Gi
          nvidia.com/gpu: "1"

```

If vllm needs to be disabled include the flag `--disable_vllm` in the container args. In this case the KServe python runtime will be used.

```yaml
apiVersion: serving.kserve.io/v1beta1
kind: InferenceService
metadata:
  name: huggingface
spec:
  predictor:
    containers:
    - args:
      - --model_name=llama2
      - --model_id=meta-llama/Llama-2-7b-chat-hf
      - --disable_vllm
      image: kserve/huggingfaceserver:latest
      name: kserve-container
      resources:
        limits:
          cpu: "6"
          memory: 24Gi
          nvidia.com/gpu: "1"
        requests:
          cpu: "6"
          memory: 24Gi
          nvidia.com/gpu: "1"
```

Perform the inference for vllm specific runtime

> vllm runtime deployments only support `/generate` endpoint for inference. Kserve python runtime supports all protocols
> Text generate Schema details - https://github.com/kserve/open-inference-protocol/blob/main/specification/protocol/generate_rest.yaml

Reviewer: ```suggestion
vllm runtime deployments only support `/generate` endpoint for inference. Please refer to [text generation API schema](https://github.com/kserve/open-inference-protocol/blob/main/specification/protocol/generate_rest.yaml) for more details.
```
